### PR TITLE
feat(dc) implement KONG_DECLARATIVE_CONFIG_STRING

### PR DESCRIPTION
**Pull request by jackkav**
### Summary

Exploring the idea of passing declarative config as a string, rather than mounting in a file.

### Full changelog

* add new env var
* add condition to parse as string rather than file
* add error handling

### Discussion

https://github.com/Kong/kong/discussions/7344

### Testing instructions

```
docker run \
-e KONG_DATABASE=off \
-p 8000:8000 \
-v $HOME/git/kong/kong/db/declarative/init.lua:/usr/local/share/lua/5.1/kong/db/declarative/init.lua \
-e KONG_DECLARATIVE_CONFIG='{"_format_version":"1.1","services":[{"host":"mockbin.com","port":443,"protocol":"https","routes":[{"paths":["/"]}]}],"plugins":[{"name":"rate-limiting","config":{"policy":"local","limit_by":"ip","minute":3}}]}' \
kong:latest
```

```
http -b :8000/status/200
http -b :8000/status/200
http -b :8000/status/200
http -b :8000/status/200
```
should see "message": "API rate limit exceeded"

### CODE REVIEW

**Comment by kikito** [comment]
I agree with the idea, I think overloading the `--config` parameter has ... drawbacks. In particular, if someone tries to load a file that doesn't exist the error they will receive will be similar to

```
could not parse the following yaml: nonnexistant.yml in `jack.json`
```

Instead of the more straightforward:
```
could not find file nonexistant.yml
```

I think (and would support) adding a separate parameter to `kong config --config-string` which explicitly parses a string, leaving the `--config` option as it is.

**Comment by gszr** [comment]
> I think (and would support) adding a separate parameter to kong config --config-string which explicitly parses a string, leaving the --config option as it is.

Agreed.

@jackkav Let us know if you need any help making this change and adding tests! Thanks for opening the PR!

File: kong/init.lua [hunk 1]
```diff
@@ -342,14 +342,20 @@ local function parse_declarative_config(kong_config)
...
-  local entities, err, _, meta = dc:parse_file(kong_config.declarative_config)
+  local entities, err, meta
+  if kong_config.declarative_config ~= nil then
+    entities, err, _, meta = dc:parse_file(kong_config.declarative_config)
+  elseif kong_config.declarative_config_string ~= nil then
+    entities, err, _, meta = dc:parse_string(kong_config.declarative_config_string)
+  end
+
   if not entities then
     return nil, "error parsing declarative config file " ..
```
Comment (gszr) [inline kong/init.lua:360]: Will also need to adjust the error message here

**Comment by jackkav** [comment]
We have taken a new approach, to add a new env var DECLARATIVE_CONFIG_STRING. 

This is in light of both, declarative config being a path in both env var and kong.conf, and the overloaded definition of conf,

```sh
docker run \
-e KONG_DATABASE=off -e KONG_LOG_LEVEL=debug \
-e KONG_DECLARATIVE_CONFIG_STRING='{"_format_version":"1.1","services":[{"host":"mockbin.com","port":443,"protocol":"https","routes":[{"paths":["/"]}]}],"plugins":[{"name":"rate-limiting","config":{"policy":"local","limit_by":"ip","minute":3}}]}' \
-p 8000:8000 \
-v $HOME/git/kong/kong/init.lua:/usr/local/share/lua/5.1/kong/init.lua \
-v $HOME/git/kong/kong/templates/kong_defaults.lua:/usr/local/share/lua/5.1/kong/templates/kong_defaults.lua \
kong:latest
```

File: kong/init.lua (hunk 1 above)
Comment (jackkav) [reply kong/init.lua]: done, in a flat way, not sure if it would be better to nest the error checking inside the first two conditions.

File: kong/init.lua [hunk 2]
```diff
@@ -342,17 +342,28 @@ local function parse_declarative_config(kong_config)
...
     -- return an empty configuration,
     -- including only the default workspace
     local entities, _, _, meta = dc:parse_table({ _format_version = "2.1" })
     return entities, nil, meta
   end
 
-  local entities, err, _, meta = dc:parse_file(kong_config.declarative_config)
+  local entities, err, meta
+  if kong_config.declarative_config ~= nil then
+    entities, err, _, meta = dc:parse_file(kong_config.declarative_config)
```
Comment (kikito) [inline kong/init.lua:354]: The linter is complaining about usage of `_` without being declared. Please add it back to the `local ` declaration on line 352

**Review by kikito — CHANGES_REQUESTED** [review]

File: spec/02-integration/02-cmd/02-start_stop_spec.lua [hunk 3]
```diff
@@ -342,6 +342,50 @@ describe("kong start/stop #" .. strategy, function()
...
+          if proxy_client then
+            proxy_client:close()
+          end
+        end)
+
+        assert(helpers.start_kong({
+          database = "off",
+          declarative_config = nil,
+          declarative_config_string = config_string,
+          nginx_worker_processes = 100, -- stress test initialization
```
Comment (kikito) [inline spec/02-integration/02-cmd/02-start_stop_spec.lua:385]: I don't think this would "stress" kong much.

The "loading of the string" happens on the init phase, and that phase is run on a single thread. When workers are created afterwards, they "copy" the Lua state which was created on the init phase. This is usually very fast.

A more appropriate way to stress kong out would be to continuously update the declarative config via simultaneous calls to `kong reload` and `POST`s requests to the `/config` Admin API endpoint. There are locks in place (the `try_lock` calls) which should prevent kong from being overrun. It should just "reject updates while it is already updating".

But that is out of the scope of this PR I think.

TL;DR: I think we should just remove this line

```suggestion
```

**Review by kikito — CHANGES_REQUESTED** [review]

**Review by kikito — APPROVED** [review]
